lib/df-detection.c

```diff
@@ -0,0 +1,242 @@
+		if (panic_on_warn) {
+			panic_on_warn = 0;
+			panic("panic_on_warn set. \n");
+		}
+	}
+}
+
+static depot_stack_handle_t dfetch_save_stack(gfp_t flags)
+{
+	unsigned long entries[STACK_DEPTH];
+	unsigned int nr_entries;
+
+	nr_entries = stack_trace_save(entries, ARRAY_SIZE(entries), 0);
+	nr_entries = filter_irq_stacks(entries, nr_entries);
+	return stack_depot_save(entries, nr_entries, flags);
+}
+
+void dfetch_add_address(const void *addr, size_t len, unsigned long caller,
+			void *kernel_addr)
+{
+	if (!in_task() || current->dfetch_addresses == NULL ||
+	    current->dfetch_pairs == NULL)
+		return;
+	if (len > MAX_LEN || addr == 0 || addr > TASK_SIZE)
+		return;
+	if (current->num_read >= current->ranges_size &&
+	    current->ranges_size < DFETCH_MAX_RECORDS) {
+		struct dfetch_address_range *temp =
+		    (struct dfetch_address_range *)krealloc(
+			current->dfetch_addresses,
+			current->ranges_size * 2 *
+			    sizeof(struct dfetch_address_range),
+			GFP_KERNEL);
+		current->dfetch_addresses =
+		    temp ? temp : current->dfetch_addresses;
+		current->ranges_size = current->dfetch_addresses
+					   ? current->ranges_size * 2
+					   : current->ranges_size;
+	}
+	if (current->num_read < current->ranges_size) {
+		current->dfetch_addresses[current->num_read].start_address =
+		    addr;
+		current->dfetch_addresses[current->num_read].len = len;
+		current->dfetch_addresses[current->num_read].caller = caller;
+		current->dfetch_addresses[current->num_read].stack =
+		    dfetch_save_stack(GFP_NOWAIT);
+		detect_intersection(kernel_addr);
+		current->num_read++;
+	}
+}
+EXPORT_SYMBOL(dfetch_add_address);
+
+void dfetch_start_system_call(void)
+{
+	if (!current->dfetch_enable)
+		return;
+	current->dfetch_addresses =
+	    (struct dfetch_address_range *)kmalloc_array(
+		DFETCH_INIT_SIZE, sizeof(struct dfetch_address_range),
+		GFP_KERNEL);
+	current->ranges_size = current->dfetch_addresses ? DFETCH_INIT_SIZE : 0;
+	current->num_read = 0;
+	current->dfetch_pairs = (struct dfetch_pair *)kmalloc_array(
+	    DFETCH_INIT_SIZE, sizeof(struct dfetch_pair), GFP_KERNEL);
+	current->dfetch_size = current->dfetch_pairs ? DFETCH_INIT_SIZE : 0;
+	current->dfetch_index = 0;
+}
+EXPORT_SYMBOL(dfetch_start_system_call);
+
+void dfetch_end_system_call(void)
+{
+	if (!current->dfetch_enable)
+		return;
+	if (current->dfetch_pairs != NULL) {
+		if (current->dfetch_index)
+			report();
+		kfree(current->dfetch_pairs);
+		current->dfetch_pairs = NULL;
+		current->dfetch_index = 0;
+		current->dfetch_size = 0;
+	}
+	if (current->dfetch_addresses != NULL) {
+		current->num_read = 0;
+		current->ranges_size = 0;
+		kfree(current->dfetch_addresses);
+		current->dfetch_addresses = NULL;
+	}
+}
+EXPORT_SYMBOL(dfetch_end_system_call);
+
+static long dfetch_ioctl(struct file *filep, unsigned int cmd,
+			 unsigned long unused)
+{
+	switch (cmd) {
+	case DFETCH_ENABLE:
+		/* Enable DF for the current task.*/
```

Comment:
You can probably move this comment in front of the function declaration and write "Enable/disable" instead of just "Enable"